lib/internal/streams/readable.js

```diff
@@ -1057,6 +1058,30 @@ Readable.prototype.wrap = function(stream) {
   return this;
 };
 
+Readable.prototype.collect = async function({
+  limit = Infinity,
+  rejectAtLimit = true
+} = {}) {
+  const bufs = [];
```

Comment:
> I would use a linked list instead of an array.

We'd then have to collapse that into the result. For object mode streams, that means an array, akin to the proposed [`%AsyncIterator.prototype%.toArray()`](https://tc39.es/proposal-iterator-helpers/#sec-asynciteratorprototype.toarray). For Buffer streams we'd have to do some other manual concatenating. For Buffer streams with encoding set, see below.

Is that collapsing going to be an OK tradeoff for the benefit of using a linked list?

> More importantly we can have a more performant implementation if we accumulate as a string.

Sure, if it's a Buffer stream with encoding set. That certainly doesn't apply to object mode streams, and I don't think it applies to Buffer streams without encoding set either?

Reply:
The reason why using a Linked List is better for this is that the result could be quite big and you'd have a LOT of memory copying to do to crease a single buffer with all the data. In essence, we'd need something like https://www.npmjs.com/package/bl to avoid all that extra performance cost.

> Sure, if it's a Buffer stream with encoding set. That certainly doesn't apply to object mode streams, and I don't think it applies to Buffer streams without encoding set either?

This analysis is correct.